src/repr/src/adt/mz_acl_item.rs

```diff
@@ -371,6 +376,146 @@ impl Columnation for MzAclItem {
+// types to and from binary, so we invent our own encoding. The encoding matches the in memory
+// format that PostgreSQL uses, which is (Oid, Oid, AclMode).
+impl AclItem {
+    pub fn encode_binary(&self) -> Vec<u8> {
+        let mut res = Vec::with_capacity(Self::binary_size());
+        res.extend_from_slice(&self.grantee.0.to_le_bytes());
+        res.extend_from_slice(&self.grantor.0.to_le_bytes());
+        res.extend_from_slice(&self.acl_mode.bits().to_le_bytes());
+        res
+    }
+
+    pub fn decode_binary(raw: &[u8]) -> Result<AclItem, Error> {
+        if raw.len() != AclItem::binary_size() {
+            return Err(anyhow!(
+                "invalid binary size, expecting {}, found {}",
+                AclItem::binary_size(),
+                raw.len()
+            ));
+        }
+
+        let oid_size = size_of::<u32>();
+
+        let grantee = Oid(u32::from_le_bytes(raw[0..oid_size].try_into()?));
+        let raw = &raw[oid_size..];
+        let grantor = Oid(u32::from_le_bytes(raw[0..oid_size].try_into()?));
+        let raw = &raw[oid_size..];
+        let acl_mode = u64::from_le_bytes(raw.try_into()?);
+
+        Ok(AclItem {
+            grantee,
+            grantor,
+            acl_mode: AclMode { bits: acl_mode },
+        })
+    }
+
+    pub const fn binary_size() -> usize {
+        size_of::<u32>() + size_of::<u32>() + size_of::<u64>()
+    }
+}
+
+impl FromStr for AclItem {
+    type Err = Error;
+
+    /// For the PostgreSQL implementation see:
+    ///   * <https://github.com/postgres/postgres/blob/9089287aa037fdecb5a52cec1926e5ae9569e9f9/src/backend/utils/adt/acl.c#L580-L609>
+    ///   * <https://github.com/postgres/postgres/blob/9089287aa037fdecb5a52cec1926e5ae9569e9f9/src/backend/utils/adt/acl.c#L223-L390>
+    ///   * <https://github.com/postgres/postgres/blob/9089287aa037fdecb5a52cec1926e5ae9569e9f9/src/backend/utils/adt/acl.c#L127-L187>
+    ///
+    /// PostgreSQL is able to look up OIDs in the catalog while parsing, so it accepts aclitems that
+    /// contain role OIDs or role names. We have no way to access the catalog here, so we can only
+    /// accept oids. Therefore our implementation is much simpler than PostgreSQL's.
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = s.split('=').collect();
+        let &[grantee, rest] = parts.as_slice() else {
+            return Err(anyhow!("invalid aclitem '{s}'"));
+        };
+
+        let parts: Vec<_> = rest.split('/').collect();
+        let &[acl_mode, grantor] = parts.as_slice() else {
+            return Err(anyhow!("invalid mz_aclitem '{s}'"));
+        };
+
+        let grantee: Oid = if grantee.is_empty() {
+            PUBLIC_ROLE_OID
+        } else {
+            Oid(grantee.parse()?)
+        };
+        let acl_mode: AclMode = acl_mode.parse()?;
+        let grantor: Oid = Oid(grantor.parse()?);
+
+        Ok(AclItem {
+            grantee,
+            grantor,
+            acl_mode,
+        })
+    }
```

Comment:
According to coverage data, this is covered by unit tests but not by end-to-end tests. Can we add some further tests to cover this? (Ignore if you feel that this does not make sense in this case.)

Reply:
There's no way to construct an aclitem in SQL yet, so we can't test this in end to end tests yet.